File: sdk/node/mcp.js
/**
 * MCP API — Model Context Protocol tool discovery and calling.
 *
 * Allows agents and external LLMs to discover and call all available
 * tools dynamically through the MCP standard.
 */

class MCPAPI {
  /** @param {import('./client').StudExClient} client */
  constructor(client) {
    this._client = client;
  }

  // ── Discovery ─────────────────────────────────────────────────────────

  /** Get MCP protocol info. */
  info() {
    return this._client._get('/mcp');
  }

  /** Discover all available MCP tools. */
  async listTools() {
    const data = await this._client._get('/mcp/tools');
    return data.tools || [];
  }

  // ── Execution ─────────────────────────────────────────────────────────

  /** Call a tool by name. */
  callTool(name, args = {}) {
    return this._client._post('/mcp/call', { name, arguments: args });
  }

  // ── Convenience wrappers ──────────────────────────────────────────────

  /** Chat with an agent via MCP. */
  async chatWithAgent(agentName, message) {
    const result = await this.callTool(`agent_${agentName}_chat`, { message });
    return result.result?.response || '';
  }

  /** Route a multi-agent task via MCP. */
  routeTask(taskDescription) {
    return this.callTool('orchestrate_task', { task: taskDescription });
  }

  /** Spawn a sandbox via MCP. */
  sandboxSpawn(template = 'python', ttlMs = 120000) {
    return this.callTool('sandbox_spawn', { template, ttlMs });
  }

  /** Run code in a sandbox via MCP. */
  sandboxRun(sandboxId, code, language) {
    const args = { sandboxId, code };
    if (language) args.language = language;
    return this.callTool('sandbox_run', args);
  }

  /** Read from the shared blackboard via MCP. */
  async blackboardGet(key) {
    const result = await this.callTool('blackboard_get', { key });
    return result.result ?? null;
  }

  /** Write to the shared blackboard via MCP. */
  blackboardSet(key, value) {
    return this.callTool('blackboard_set', { key, value });
  }
}

module.exports = MCPAPI;

File: sdk/python/studex_sdk/mcp.py
"""
MCP API — Model Context Protocol tool discovery and calling.

Allows agents and external LLMs to discover and call all available
tools dynamically through the MCP standard.
"""


class MCPAPI:
    """
    Wrapper around the StudEx MCP (Model Context Protocol) API.

    Usage:
        client = StudExClient()
        tools = client.mcp.list_tools()
        result = client.mcp.call_tool("sandbox_spawn", {"template": "python"})
    """

    def __init__(self, client):
        self._client = client

    # ── Discovery ─────────────────────────────────────────────────────────

    def info(self):
        """
        Get MCP protocol info.

        Returns:
            dict with keys: protocol, version, endpoints, toolsCount, agentsCount
        """
        return self._client._get("/mcp")

    def list_tools(self):
        """
        Discover all available MCP tools.

        Returns:
            list of tool dicts with keys: name, description, inputSchema
        """
        data = self._client._get("/mcp/tools")
        return data.get("tools", [])

    # ── Execution ─────────────────────────────────────────────────────────

    def call_tool(self, name, arguments=None):
        """
        Call a tool by name.

        Args:
            name: Tool name (e.g. "sandbox_spawn", "agent_cashclaw_chat")
            arguments: dict of parameters matching the tool's inputSchema

        Returns:
            Tool-specific result dict with keys: tool, status, result
        """
        return self._client._post("/mcp/call", {
            "name": name,
            "arguments": arguments or {},
        })

    # ── Convenience wrappers ──────────────────────────────────────────────

    def chat_with_agent(self, agent_name, message):
        """
        Convenience: chat with an agent via MCP.

        Args:
            agent_name: One of "cashclaw", "hermes", "openhuman", "cursor", "farm"
            message: Message to send

        Returns:
            Agent response text
        """
        result = self.call_tool(f"agent_{agent_name}_chat", {"message": message})
        return result.get("result", {}).get("response", "")

    def route_task(self, task_description):
        """
        Convenience: route a multi-agent task via MCP.

        Args:
            task_description: e.g. "deploy the dashboard"

        Returns:
            Task routing result
        """
        return self.call_tool("orchestrate_task", {"task": task_description})

    def sandbox_spawn(self, template="python", ttl_ms=120_000):
        """
        Convenience: spawn a sandbox via MCP.

        Args:
            template: "python", "node", "shell", or "go"
            ttl_ms: Time-to-live in ms

        Returns:
            dict with keys: id, template, workDir, commands
        """
        return self.call_tool("sandbox_spawn", {
            "template": template,
            "ttlMs": ttl_ms,
        })

    def sandbox_run(self, sandbox_id, code, language=None):
        """
        Convenience: run code in a sandbox via MCP.

        Args:
            sandbox_id: Sandbox ID
            code: Source code to run
            language: Optional language override

        Returns:
            dict with keys: stdout, stderr, exitCode
        """
        args = {"sandboxId": sandbox_id, "code": code}
        if language:
            args["language"] = language
        return self.call_tool("sandbox_run", args)

    def blackboard_get(self, key):
        """
        Convenience: read from the shared blackboard via MCP.

        Args:
            key: Fact key

        Returns:
            Fact value or None
        """
        result = self.call_tool("blackboard_get", {"key": key})
        return result.get("result")

    def blackboard_set(self, key, value):
        """
        Convenience: write to the shared blackboard via MCP.

        Args:
            key: Fact key
            value: Any JSON-serializable value
        """
        return self.call_tool("blackboard_set", {"key": key, "value": value})

    def __repr__(self):
        return "<MCPAPI>"
